### PR TITLE
feat(epic): cap summarise().skippedDetail + single-pass reducer

### LINKED ISSUE #1120: epic-connector: cap skippedDetail array size in BullMQ job result payload
## Context
Identified during review of PR #1117.

## Description
`summarise()` now includes `skippedDetail: SkippedDetail[]` (an unbounded array spread per resource type) in the BullMQ job return value. `job.returnvalue` is persisted to Redis. Today this is at most ~5 entries (Observation has 2 categories, MedicationRequest 1, others zero), but #1114 (multi-status MedicationRequest fan-out) and future per-status / per-category fan-outs could grow this materially.

Suggested cap: surface up to N (e.g. 50) full entries in `skippedDetail`, then a `skippedTruncated: number` counter. The full set is already persisted to `epic_sync_state.skipped_sub_resources` so operators have a recovery path via the DB if they need the long tail.

Not blocking today — flagging so we don't bloat job-result payloads as fan-out grows.

## Acceptance Criteria
- [ ] summarise() caps skippedDetail.length to a constant (e.g. 50)
- [ ] When truncated, the rollup includes a `skippedDetailTruncated: number` counter so dashboards can render \"+N more\"
- [ ] DB-persisted skipped_sub_resources remains uncapped (the full set still lands in epic_sync_state via markOk/markFailed)

### LINKED ISSUE #1121: epic-connector: avoid O(N^2) array-spread in summarise() reducer
## Context
Identified during review of PR #1117.

## Description
`summarise()` builds `skippedDetail` via `[...acc.skippedDetail, ...r.skipped]` inside a `reduce`. For each iteration this allocates a new array and copies all prior entries — O(N²) where N is the total skipped count. Today N <= ~5 so it's noise, but this is a code-smell pattern that will bite as fan-out grows (especially combined with #1114 multi-status fan-out).

Trivial fix: replace with a single `flatMap(r => r.skipped)` outside the reduce, or push into a pre-allocated array.

## Acceptance Criteria
- [ ] summarise() uses flatMap or a single-pass push instead of spread-in-reduce
- [ ] Existing summarise tests still pass unchanged (behaviour-preserving refactor)

### PR DESCRIPTION
**Pull request by blamechris**
## Summary
Bundles two PR #1117 follow-ups around `summarise()` in `sync-worker.ts`. Both touch the same function and integrate naturally — capping needs the reducer to track an overflow counter, single-pass makes the cap check trivial.

- **#1120** — Cap `skippedDetail` at `SUMMARISE_SKIPPED_DETAIL_CAP = 50` so the Redis-persisted BullMQ job result payload stays bounded as fan-out widens. Overflow reported via new `skippedDetailTruncated: number` field; dashboards can render \"+N more\". Full skipped set is still persisted to `epic_sync_state.skipped_sub_resources` via markOk/markFailed — operators have a DB recovery path for the long tail.
- **#1121** — Replaced `[...acc.skippedDetail, ...r.skipped]` spread-in-reduce with a single-pass for-of loop. Old pattern was O(N²) (fresh array allocation + full copy per iteration). Not measurable today with N ≤ 5 but matters as `SUPPORTED_RESOURCE_TYPES` grows, especially combined with the future multi-status fan-out under #1114.

Closes #1120.
Closes #1121.

## API change
- `SyncSummary` adds `skippedDetailTruncated: number` (additive, defaults to 0)
- `SUMMARISE_SKIPPED_DETAIL_CAP` exported so admin tooling can reason about the cap

## Why 50
Today's max fan-out produces ≤5 skipped entries per sync (2 Observation categories + 1 MedicationRequest status). 50 leaves 10× headroom even if every supported resource type starts fanning out, and keeps the typical BullMQ payload at <5KB. If #1114 (multi-status MedicationRequest) ships and pushes the typical case past 20, we can tune up; not worth making it env-configurable yet.

## Cap semantics
- `skipped` (count) — always the true total, regardless of cap
- `skippedDetail` (entries) — first N entries by iteration order, capped at 50
- `skippedDetailTruncated` (overflow count) — `skipped - skippedDetail.length`

Deterministic order matters: the cap preserves the FIRST N entries, not a random sample, so the truncated payload is reproducible across re-runs of the same sync.

## Test plan
- [x] `pnpm --filter @carebridge/epic-connector test` — 205 pass + 1 documented skip (was 201, +4 new)
  - New: \"exposes SUMMARISE_SKIPPED_DETAIL_CAP for callers that need the cap value\"
  - New: \"skippedDetail count == skipped count when under the cap; truncation counter stays at 0\"
  - New: \"caps skippedDetail at CAP and reports the truncated count\" (CAP+17 entries → 50 in detail, 17 in truncated)
  - New: \"cap applies across multiple SyncResult entries\" (truncation kicks in mid-iteration, verifies reducer respects cap across boundaries)
  - Existing 5 summarise tests still pass; updated empty-input test to assert `skippedDetailTruncated: 0`
- [x] `pnpm typecheck` — workspace 38/38 clean

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #1127

### Summary
Bundles two #1117 follow-ups (#1120 cap + #1121 single-pass) that touch the same `summarise()` function — natural fit since the cap check is trivial in a single-pass loop. Additive `skippedDetailTruncated` field, deterministic cap order, full skipped set still persisted to `epic_sync_state` via markOk/markFailed. 9 tests in the file (5 existing + 4 new), assertions match the PR description. Approve.

### Strengths
- Single-pass reducer is a clean, idiomatic O(N) replacement for the spread-in-reduce O(N²) pattern; no behavior change for the existing 5 assertions.
- Cap semantics are tight and well-documented in JSDoc: `skipped` stays the true total, `skippedDetail` keeps first N, `skippedDetailTruncated = skipped - skippedDetail.length`.
- `skippedDetailTruncated` defaults to 0 — additive, non-breaking for any existing dashboard reading `job.returnvalue`.
- The CAP+17 test uses `SUMMARISE_SKIPPED_DETAIL_CAP` symbolically; bumping the cap later doesn't break the test.
- "Cap applies across multiple SyncResult entries" test (mid-iteration truncation) is the right edge case for a reducer with global state.
- Recovery path (full skipped set on `epic_sync_state.skipped_sub_resources`) is explicitly called out in the cap's JSDoc — operators know where to look for the long tail.

### Issues Found

#### Critical (Must Fix)
(none)

#### Suggestions (Should Consider)
| File | Line | Suggestion | Rationale |
|------|------|------------|-----------|
| `services/epic-connector/src/index.ts` | export block | Re-export `SUMMARISE_SKIPPED_DETAIL_CAP` and the `SyncSummary` type from package root | Cap is currently only reachable via deep subpath `workers/sync-worker.js`. `SyncSummary` is an unexported `interface` — any consumer typing `job.returnvalue` redefines the shape and drifts when fields are added (just happened with `skippedDetailTruncated`). Tracked: #1128. |
| `services/epic-connector/src/workers/sync-worker.ts` | 217-225, 227-241 | Document that cap preserves `SyncResult[]` iteration order as a contract, not just an implementation detail; cross-reference #1114 in cap JSDoc | Test asserts first-N preservation, but caller order is implicit. After #1114 ships, typical fan-out could approach the cap and a transient auth flap could fill it. Tracked: #1129. |

#### Nitpicks (Optional)
- `skippedDetailTruncated` naming reads slightly oddly next to `skippedDetail` (the field name is the past-participle "truncated" carrying a count). `skippedDetailDroppedCount` or `skippedDetailOverflow` would be more self-describing. Bikeshed — defer.

### Deferred Items (Follow-Up Issues)
| Suggestion | Issue | Rationale for deferral |
|------------|-------|------------------------|
| Re-export `SUMMARISE_SKIPPED_DETAIL_CAP` + `SyncSummary` from package root | #1128 | Pure additive surface change; can ship independently when the first external consumer materializes. |
| Document iteration-order stability + revisit cap after #1114 | #1129 | Docs-only nudge today; the real re-evaluation happens when multi-status fan-out lands. |

### Architecture Notes
The cap defends the Redis-persisted BullMQ `job.returnvalue` payload, which is the right boundary to protect — Redis memory is the scarce resource, not Postgres. Keeping the full skipped set on `epic_sync_state.skipped_sub_resources` preserves operator recovery without bloating the queue. Single-pass refactor doesn't change semantics (no array-spread, but iteration order is identical), so the diff is a behavior-preserving perf win plus an additive observability field. The only structural smell is that `SyncSummary` describes the public BullMQ result shape but isn't actually exported as a type — that's the #1128 follow-up, not a blocker.

### Verdict
- [x] Approve
- [ ] Request Changes
- [ ] Comment

File: services/epic-connector/src/workers/sync-worker.ts [hunk 1]
```diff
@@ -208,37 +208,68 @@ interface SyncSummary {
...
+    conflicts += r.conflicts;
+    errors += r.errors.length;
+    skipped += r.skipped.length;
+    for (const s of r.skipped) {
+      if (skippedDetail.length < SUMMARISE_SKIPPED_DETAIL_CAP) {
+        skippedDetail.push(s);
+      } else {
+        skippedDetailTruncated++;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (blamechris) [reply services/epic-connector/src/workers/sync-worker.ts]: **FIX**

Fixed in `5c9fcf0`

**Change:** Compute headroom upfront and short-circuit the iteration past the cap. When this SyncResult's skipped length fits in the remaining headroom, push it in one loop. When the cap is already full, bump `skippedDetailTruncated` by `r.skipped.length` and continue.

```diff
  for (const r of results) {
    imported += r.imported;
    updated += r.updated;
    conflicts += r.conflicts;
    errors += r.errors.length;
    skipped += r.skipped.length;
-   for (const s of r.skipped) {
-     if (skippedDetail.length < SUMMARISE_SKIPPED_DETAIL_CAP) {
-       skippedDetail.push(s);
-     } else {
-       skippedDetailTruncated++;
-     }
-   }
+   const headroom = SUMMARISE_SKIPPED_DETAIL_CAP - skippedDetail.length;
+   if (headroom <= 0) {
+     skippedDetailTruncated += r.skipped.length;
+     continue;
+   }
+   if (r.skipped.length <= headroom) {
+     for (const s of r.skipped) skippedDetail.push(s);
+   } else {
+     for (let i = 0; i < headroom; i++) skippedDetail.push(r.skipped[i]!);
+     skippedDetailTruncated += r.skipped.length - headroom;
+   }
  }
```

Matters for the future #1114 multi-status MedicationRequest fan-out where a single SyncResult could carry hundreds of skipped entries — without the short-circuit we'd iterate every one of them past the cap.

Existing 9 tests pass unchanged (behaviour-preserving).

**Comment by blamechris** [comment]
## Review Comments Addressed

| # | Comment | Outcome | Reference |
|---|---------|---------|-----------|
| 1 | Iterating every overflow entry just to increment the truncated counter wastes O(N) work past the cap | FIX | `5c9fcf0` |

**Total:** 1 inline comment addressed
- Fixed: 1 (commit `5c9fcf0`)
- False positives: 0
- Deferred follow-up issues: 2 (#1128 re-export SUMMARISE_SKIPPED_DETAIL_CAP + SyncSummary from package root; #1129 document iteration-order stability + revisit cap after #1114)
